src/net/http.rs

```diff
@@ -3747,36 +3775,141 @@ impl HttpResponseType {
+        // a page ID.  Expect end-of-stream after this.
+        retry_reader.set_position(pos);
+        let next_page: Txid = match read_next(retry_reader) {
+            Ok(txid) => txid,
+            Err(e) => match e {
+                codec_error::ReadError(ref ioe) => match ioe.kind() {
+                    io::ErrorKind::UnexpectedEof => {
+                        if pos == retry_reader.position() {
+                            // this is fine -- the node didn't get another page
+                            return Ok(None);
+                        } else {
+                            // partial data -- corrupt stream
+                            test_debug!("Unexpected EOF: {} != {}", pos, retry_reader.position());
+                            return Err(e.into());
+                        }
+                    }
+                    _ => {
+                        return Err(e.into());
+                    }
+                },
+                e => {
+                    return Err(e.into());
+                }
+            },
+        };
+
+        test_debug!("Read page_id {:?}", &next_page);
+        Ok(Some(next_page))
+    }
+
+    /// Decode a transaction stream, returned from /v2/mempool/query.
+    /// The wire format is a list of transactions (no SIP-003 length prefix), followed by an
+    /// optional 32-byte page ID.  Obtain both the transactions and page ID, if it exists.
+    pub fn decode_tx_stream<R: Read>(
         fd: &mut R,
         len_hint: Option<usize>,
-    ) -> Result<HttpResponseType, net_error> {
-        // NOTE: there will be no length prefix on this
-        let mut txs = vec![];
+    ) -> Result<(Vec<StacksTransaction>, Option<Txid>), net_error> {
+        // The wire format is `tx, tx, tx, tx, .., tx, txid`.
+        // The last 32 bytes are the page ID for the next mempool query.
+        // NOTE: there will be no length prefix on this.
+        let mut txs: Vec<StacksTransaction> = vec![];
         let max_len = len_hint.unwrap_or(MAX_MESSAGE_LEN as usize) as u64;
         let mut bound_reader = BoundReader::from_reader(fd, max_len);
+        let mut retry_reader = RetryReader::new(&mut bound_reader);
+        let mut page_id = None;
+        let mut expect_eof = false;
+
         loop {
-            let tx: StacksTransaction = match read_next(&mut bound_reader) {
-                Ok(tx) => Ok(tx),
+            let pos = retry_reader.position();
+            let next_msg: Result<StacksTransaction, _> = read_next(&mut retry_reader);
+            match next_msg {
+                Ok(tx) => {
+                    if expect_eof {
+                        // this should have failed
+                        test_debug!("Expected EOF; got transaction {}", tx.txid());
+                        return Err(net_error::DeserializeError(
+                            "Expected end-of-stream".to_string(),
+                        ));
```

Comment:
Would it make sense to add a `net_error::ExpectedEndOfStream` variant? This same error string is produced three times here.

Reply:
Sure, I can try adding that.  I don't *think* anything will break, but we'll see.